### DISCUSSION TITLE
json object in condition doesn't utilize postgres GIN index due to ->> operator use instead of @>, @<, ? or jsonb_path_exists()

### DISCUSSION THREAD

**Discussion by dginzbourg** [mikro-orm/mikro-orm#4465] [General]
Hi
Initially I used the default json syntax in condition, but it lead to bad query performance. I'm opening this discussion for othesrs to be aware of the issue and maybe add it as a feature request (or bug fix).

The raw query was translated to something like:
```sql
select * from event
where raw_event->>property1 = 's1'
```
It's postgres db, raw_event is jsonb type, there's GIN index on raw_event field. The performance of the query was bad. The execution plan showed that index wasn't used at all. Our DBA found out that different operator has to be used to utilize the index:
```sql
select * from event
where raw_event @> '{ "property1": "s1" }'
```
If the condition is looking for absence of 'property1' then there're two options (assuming property1 doesn't exist in the record with null value):
```sql
select * from event
where raw_event ? 'property1'
```
or
```sql
select * from event
where jsonb_path_exists(raw_event, '$.property1') = false
-- you can use also not jsonb_path_exists(raw_event, '$.property1')
```

Is there a reason why Mikroorm doesn't use those operators when json object is in a condition? 
I ended up with using something like:
```js
     ...
 .where({ [expr(`jsonb_path_exists(raw_event, '$.property1')`)]: false })
 .orWhere({ [expr('raw_event @> \'{ "property1": "s1"}\'')]: [] })
```
But it's cumbersome. This "syntax" doesn't support "IN" (could be a nice addition for IN translation to OR of json's).
There's also a problem with `where raw_event ? 'property1'` syntax, expr() didn't help, seems there's still a bug in knex that doesn't allow to escape '?' (I failed to understand why I'm getting this error, fortunately was able to use jsonb_path_exists() instead) 
I tried :
```sql
.where({ [expr(`raw_event \\? 'property1')`)]: [] })
```
but got an error and :
```
select "ent"."id" from "aentity" as "ent" where raw_event ? 'property1' - syntax error at or near "\"
error: select "ent"."id" from "aentity" as "ent" where raw_event ? 'property1' - syntax error at or near "\"
    at Parser.parseErrorMessage (/Users/dginzbourg/repos/vcp-bulk/node_modules/pg-protocol/src/parser.ts:369:69)
    at Parser.handlePacket (/Users/dginzbourg/repos/vcp-bulk/node_modules/pg-protocol/src/parser.ts:188:21)
    at Parser.parse (/Users/dginzbourg/repos/vcp-bulk/node_modules/pg-protocol/src/parser.ts:103:30)
    at Socket.<anonymous> (/Users/dginzbourg/repos/vcp-bulk/node_modules/pg-protocol/src/index.ts:7:48)
    at Socket.emit (node:events:513:28)
    at addChunk (node:internal/streams/readable:324:12)
    at readableAddChunk (node:internal/streams/readable:297:9)
    at Socket.Readable.push (node:internal/streams/readable:234:10)
    at TCP.onStreamRead (node:internal/stream_base_commons:190:23)
    at TCP.callbackTrampoline (node:internal/async_hooks:130:17)
```

**Comment by B4nan** [comment]
> Is there a reason why Mikroorm doesn't use those operators when json object is in a condition?

Because we need a way to support all possible operators, and this wouldn't work at all with anything else than a simple comparison, you said it yourself, even `IN` query is not supported, which is crucial, same for `LIKE` queries, etc.

To be honest, this all sounds more like a postgres bug (not using indexes when it should), are you sure it's not just about that? Are you using latest postgres? Have you checked their bug reports?

Also, you haven't shared a single snippet about how you actually used the ORM initially, only the workarounds with QB. I guess it all boils down to that being too generic, and you expect too tailored queries for your use case. 

Lastly, there is no shame in using raw queries.

**Comment by dginzbourg** [comment]
We're using Aurora RDS Postgres 15.3. I checked Postgres doc and it states [The default GIN operator class for jsonb supports queries with the key-exists operators ?, ?| and ?&, the containment operator @>, and the jsonpath match operators @? and @@. ](https://www.postgresql.org/docs/current/datatype-json.html) (see `8.14.4. jsonb Indexing` section). So it isn't a bug but a "feature" :). 

I opened the discussion 1) to check whether it's a bug or favoring generic and 2) for others be aware of subtleness of using ORM for condition on jsonb field (maybe worth to mention in the documentation) and workarounds. 

I didn't include the code assuming it's something known/common. Here's the code (query builder and findOne()):
```js
@Entity()
export class AEntity {
  @PrimaryKey({ autoincrement: true })
  id!: number;
 
  @Property({ type: JsonType, columnType: 'jsonb' })
  raw_event: any;
}

export class EntityOperationLogger implements EventSubscriber {
  async afterUpdate<T>(args: EventArgs<T>): Promise<void> {
    console.log('EntityOperationLogger', args);
  }
}

describe('Issue', () => {
  let orm: MikroORM;

  beforeAll(async () => {
    orm = await MikroORM.init({
      metadataProvider: TsMorphMetadataProvider,
      // entities: [EventEntityTest],
      entities: [AEntity, BEntity],
      host: 'localhost',
      port: 5432,
      user: '***',
      password: '***',
      dbName: 'test',
      schema: 'test',
      driver: PostgreSqlDriver,
      allowGlobalContext: true,
      debug: true,
      batchSize: 1000,
      forceUtcTimezone: true,
      forceUndefined: true,
      useBatchInserts: true,
      useBatchUpdates: true,
      cache: { options: { cacheDir: 'dist' } },
      subscribers: [new EntityOperationLogger()],
    });
    await orm.schema.createSchema();
  });

  afterEach(async () => {
    await orm.schema.dropSchema();
  });

  afterAll(() => orm.close(true));

  test('json query', async () => {
    const em: EntityManager = orm.em as SqlEntityManager;
    const qb = em.createQueryBuilder<AEntity>(AEntity, 'ent', 'read');

    await orm.em.nativeInsert(AEntity, { id: 1, raw_event: { run_id: 2, application_id: 3 } });

    qb.select(['id']).where({ raw_event: { run_id: 'run1' } });
    const res = await qb.execute('all');

    await orm.em.findOne(AEntity, { raw_event: { run_id: 'run2' } });
  });
 });
```

we don't create index via ORM, not sure how to add it in the code above
```sql
create index event_raw_event on aentity using gin(raw_event);
```

the queries produced are:
```sql
select "ent"."id" from "test"."aentity" as "ent" where "ent"."raw_event"->>'run_id' = 'run1'
```
```sql
select "a0".* from "test"."aentity" as "a0" where "a0"."raw_event"->>'run_id' = 'run2'
```
